[#338] Add docs for running in production (#359)
airyhq/airy: docs/docs/guides/airy-core-in-production.md

> # Running the Airy Core Plaform in production

Reviewer: we miss the front matter here (they configuration part of the document. You can see it in any other doc

> The cloud providers part can be left as TODO in the first pr. It's enough to cover all the requirements (size of a kubernetes cluster, how to get the code up and running in the cluster, how to run the services (redis, postgres, kafka, zookeeper) both in the same kubernets cluster or how to configure the apps so that the system works)

Reviewer: I suppose those are notes?

Author: yes, will remove it

## Architecture

> The Airy Core Platform is messaging platform, based on microservices and comprised of a backend and frontend system. 

Reviewer 2: ```suggestion
The Airy Core Platform is a microservice based messaging platform comprised of a backend and frontend system. 
```

Reviewer: I wouldn't use the work microservice as we don't really buy on the philosophy much (take api-communication as an example)

Author: ok, will remove microservices

The `backend` system is a streaming platform and its role is to:
- Receive messages or events from different sources, process them on the fly and store them in Kafka `(sources)`
- Make them available and accessible through an API (api)

> - Send messages or events to external parties (webhooks)

Reviewer 2: Here I would link to `glossary.md#sources`, `http.md` and `webhook.md`

Author: Do you mean instead of this text or as an addition?

> Apart of the messaging infrastructure, there is also an `admin` app which manages users, authentication and the other features related to the organization using the Airy Core Platform and its users.

Reviewer: ```suggestion
Furthermore, the platform provides an `admin` application which manages authentication and authorization features.
```

> The `frontend` system is comprised of a webapp and a custom Airy source called the `Chat Plugin`.

Reviewer: ```suggestion
The `frontend` system is comprised of a demo application and the JavaScript integration of  the [Chat Plugin](/docs/sources/chat-plugin.md).
```

Having that in mind, the following docker containers or `Airy apps` need to run in order for the Airy Core Platform to be fully functional:

Sources:
- sources-facebook-webhook - Receive messages from Facebook
- sources-facebook-events-router - Process messages from Facebook
- sources-facebook-sender - Send messages to the Facebook source
- sources-google-webhook - Receive messages from Google
- sources-google-events-router - Process messages from Google
- sources-google-sender - Send messages to the Google source

> - sources-chatplugin - Receive and processing messages from the Airy Chat Plugin source

Reviewer: I think instead of enumerating the actual apps, I would present the convention we use (the doc will inevitably run out of data as we add sources otherwise)

Author: I thought that the name in ever bullet should link to a documentation for the particular app (docker registry, variables, ports, liveness-probe). This can also be a part of a separate `architecture` document.

For the documentation of the individual apps, Github packes isn't good enough, if we just link to the repository. So perhaps we use a separate document for every app or hub.docker.com.

Reviewer: we discussed this offline and I think we're on the same page (about conventions being more important than the specific pod)

API:
- api-auth - Authenticate and authorize users
- api-admin - Manage sources and other features
- api-communication - Send or receive messages through the API

Webhook:
- webhook-publisher - Process events to be sent to external parties and write them in a redis store
- webhook-consumer - Pull events from redis and send them to external webhooks

Frontend:
- frontend-demo - Webapp for managing messages and administration of the organization

> - frontend-chat-plugin - Airy Chat Plugin for 

Reviewer: I'd use `### API` headers for the paragraph (so they're also linkable)

Depending on implementation of the user, different choices can be made regarding the implementation of the underlying services and the whole system in general. As first pre-requisites, the underlying services are needed to be running. They can be implemented as separate docker containers, in different kubernetes clusters, on a separate virtual machines or physical servers. Here is a brief description on the implementation and how to fully achieve high availability.

### The kafka store

The Kafka store can be comprised of multiple kafka brokers, zookeepers and Schema registry servers. The Airy apps can work with a single broker, however for running in production one can argue that this isn't enough. We recommend using at least five kafka brokers, having all the topics with a setting of replica factor 3. That way, even if two brokers become unavailable - the system will be fully operational. 

We also recommend running with at least three zookeepers and two Schema registry servers.

> If deployed on kubernetes, consider running the app on a StatefulSet workload, for persistent storage.

Reviewer: which app?

---
id: airy-core-production
title: Running the Airy Core Plaform in production
sidebar_label: Airy Core Platform in production
slug: /airy-core-production
---

> This document gives a general knowledge of the architecture of the system, explaining how it should be deployed in a production environment.

Reviewer: ```suggestion
This document provides an overview of the architecture of the Airy Core Platform and good practices for production environments.
```

## Architecture

The Airy Core Platform is a messaging platform comprised of a backend and frontend system (run as containers)

The `backend` system is a streaming platform and its role is to:
- Receive messages or events from different sources, process them on the fly and store them in Kafka [sources](glossary.md#sources)

> - Make them available and accessible through an API [api](http.md)

Reviewer: ```suggestion
- Make them available and accessible through an API [api](api/http.md)
```

> - Send messages or events to external parties [webhooks](webhook.md)

Reviewer: ```suggestion
- Send messages or events to external parties [webhooks](api/webhook.md)
```

Furthermore, the platform provides an `admin` application which manages authentication and authorization features.

> The `frontend` system is comprised of a demo application and the JavaScript integration of  the [Chat Plugin](/docs/sources/chat-plugin.md).

Reviewer: ```suggestion
The `frontend` system is comprised of a demo application and the JavaScript integration of  the [Chat Plugin](sources/chat-plugin.md).
```

### API
- api-auth - Authenticate and authorize users
- api-admin - Manage sources and other features
- api-communication - Send or receive messages through the API

### Webhook
- webhook-publisher - Process events to be sent to external parties and write them in a redis store
- webhook-consumer - Pull events from redis and send them to external webhooks

### Frontend
- frontend-demo - Webapp for managing messages and administration of the organization
- frontend-chat-plugin - Airy Chat Plugin for 

## Requirements 

The Airy apps should be started as separated docker containers and they must have access to several other underlying services, which need to be in place for the system to be fully functional:
- Kafka, Zookeeper and the Schema registry: These three services comprise the kafka store and need to be in place as they are the base store for messages and events which are received and processed by the Airy Core Platform. Kafka needs Zookeeper to be in place to work and the Schema registry is an application which is the central schema repository for all the data written in kafka, used by all of the Airy apps.
- Postgres
- Redis

> The Airy apps can run on a docker runtime engine, however we recommend deploying the Airy Core Platform inside kubernetes. For bootstrapping the entire Airy Core Platform on a single machine, we have included a Vagrant configuration, inside the `infrastructure` directory, which creates a K3s kubernetes implementation, running on top of Alpine Linux. All of the underlying services and the Airy apps run inside this kubernetes box, however, they can also be deployed in a different manner, in order to fully grasp the high availability and redundancy of the system.

Reviewer: ```suggestion
The Airy apps can run on a docker runtime engine, however we recommend deploying the Airy Core Platform inside kubernetes. For bootstrapping the entire Airy Core Platform on a single machine, we included a Vagrant configuration, inside the `infrastructure` directory, which creates a K3s kubernetes implementation, running on top of Alpine Linux. All of the underlying services and the Airy apps run inside this kubernetes box, however, they can also be deployed in a different manner.
```

Depending on implementation of the user, different choices can be made regarding the implementation of the underlying services and the whole system in general. As first pre-requisites, the underlying services are needed to be running. They can be implemented as separate docker containers, in different kubernetes clusters, on a separate virtual machines or physical servers. Here is a brief description on the implementation and how to fully achieve high availability.

> ### The kafka store

Reviewer: ```suggestion
### Apache Kafka cluster
```

The Kafka store can be comprised of multiple kafka brokers, zookeepers and Schema registry servers. The Airy apps can work with a single broker, however for running in production one can argue that this isn't enough. We recommend using at least five kafka brokers, having all the topics with a setting of replica factor 3. That way, even if two brokers become unavailable - the system will be fully operational. 

We also recommend running with at least three zookeepers and two Schema registry servers.

If deployed on kubernetes, consider running Kafka and Zookeeper on a StatefulSet workload, for persistent storage. The Schema registry can be run as a Deployment, as it is stateless.

Apart from running your own kafka cluster, a managed version can be used.

> ### Postgresql

Reviewer: ```suggestion
### PostgreSQL
```

A cluster of postgres servers can be used or a managed one such as AWS RDS.
If deployed on kubernetes, consider running the app on a StatefulSet workload, for persistent storage.

### Redis

A cluster of redis servers can be used or a managed one such as AWS ElastiCache for Redis
If deployed on kubernetes, consider running the app on a StatefulSet workload, for persistent storage.

### Connecting the Airy apps to the underlying services

After the underlying services are deployed, the Airy apps need to be created as docker containers on any docker runtime or inside kubernetes. Regardless of the deployment decision, they must have network access to the underlying services through the following variables which need to be defined before starting the containers.

- KAFKA_BROKERS - a comma separated list of kafka brokers (ex. `kafka1.airy:9092,kafka2.airy:9092`)
- KAFKA_SCHEMA_REGISTRY_URL - the url of the schema registry (ex. `http://airy-cp-schema-registry:8081`)
- ZOOKEEPER_SERVERS
- POSTGRES_USER
- POSTGRES_DB
- POSTGRES_PASSWORD
- REDIS_HOSTNAME
- REDIS_PORT

> At the moment, The Airy Core Platform is fully tested on k3s, minikube and AWS EKS and there aren't any specific issues.

Reviewer: ```suggestion
At the moment, the Airy Core Platform is fully tested on k3s, minikube and AWS EKS and there aren't any specific issues.
```

---
title: Running the Airy Core Plaform in production
sidebar_label: Production
---

This document provides an overview of the architecture of the Airy Core Platform and good practices for production environments.

## Architecture

> The Airy Core Platform is a messaging platform comprised of a backend and frontend system, both of which run on top of multiple docker containers.

Author: will remove the part with `run on top of multiple docker containers`

- Make them available and accessible through an [API](api/http.md)
- Send messages or events to external parties [webhooks](api/webhook.md)

Furthermore, the platform provides an `admin` application which manages authentication and authorization features.

The `frontend` system is comprised of a demo application and the JavaScript integration of  the [Chat Plugin](sources/chat-plugin.md).

Having that in mind, the docker containers or `Airy apps` which need to run in order for the Airy Core Platform to be fully functional, have the following naming convention:

### Sources
- sources-`SOURCE_NAME`-webhook - Receive messages from a specific source
- sources-`SOURCE_NAME`-events-router - Process messages from a specific source
- sources-`SOURCE_NAME`-sender - Send messages to a specific source

### API
- api-`ENTITY` - Backend services, exposed to the outside (admin, authentication, communication, etc.)

### Webhook
- webhook-`ENTITY` - Process events to be sent to external parties

### Frontend
- frontend-`ENTITY` - Frontend application, communicating with the `API` services

## Requirements 

> The `Airy apps` should be started as separated docker containers and they must have access to several other underlying services, which need to be in place for the system to be fully functional:

Reviewer: ```suggestion
The `Airy apps` must run as separated docker containers and require access to several services which need to be in place for the system to be functional:
```

> - `Kafka cluster`: Kafka, Zookeeper and the Schema registry. These three services comprise the kafka store and need to be in place as they are the base store for messages and events which are received and processed by the Airy Core Platform. Kafka needs Zookeeper to be in place to work and the Schema registry is an application which is the central schema repository for all the data written in kafka, used by the streaming apps inside the Airy Core Platform.

Reviewer: ```suggestion
- `Kafka cluster`: Kafka, Zookeeper and the Confluent Schema registry. These three services comprise the Kafka store. They are the default storage system of the Airy Core Platform. Kafka requires Zookeeper to  work. The Confluent Schema registry facilitates Avro typed data pipe-lines. All our Kafka based applications require the registry to work.
```

> - `PostgreSQL`: Holding a database which stores organization and user related data.

Reviewer: ```suggestion
- `PostgreSQL`: Where we store authentication data.
```

> - `Redis`: Data structure used as a simple message queue, for events which are sent to external webhooks.

Reviewer: ```suggestion
- `Redis`: The queuing system used by our webhook system relies on Redis.
```

> Apart from the underlying services, a docker runtime or a running Kubernetes cluster must also be available, for deploying the `Airy apps`.

Reviewer: ```suggestion
A docker runtime or a running Kubernetes cluster must also be available to deploy the `Airy apps`.
```

Reviewer: right in the next paragraph we say it can be run anywhere... so this comes across as confusing. I think it's ok we require Kubernetes for now (it's also simpler to deploy so many services right now this way)

Author: On `any` infrastructure, so I think that is still accurate. One can run all of  the base services and kubernets or docker engine on his own infrastructure.

But maybe instead of  on `any infrastructure` I can put `on your own infrastructure`? But `any` also indicates that it can be run in your data center or your custom hosting provider...

## Running on any infrastructure

### Base infrastructure

Depending on implementation of the user, different choices can be made regarding the implementation of the underlying services and the whole system in general. As first pre-requisites, the underlying services are needed to be running. They can be implemented as separate docker containers, in different Kubernetes clusters, on a separate virtual machines or physical servers. Here is a brief description on the implementation and how to fully achieve high availability.

### The Kafka cluster

> The Kafka store can be comprised of multiple kafka brokers, zookeepers and Schema registry servers. The Airy apps can work with a single broker, however for running in production one can argue that this isn't enough. We recommend using at least five kafka brokers, having all the topics with a setting of replica factor 3. That way, even if two brokers become unavailable - the system will be fully operational. 

Reviewer: ```suggestion
The Kafka store requires multiple Kafka brokers, zookeepers, and Confluent Schema registry servers. We recommend using at least five Kafka brokers and  set the replication factor of _all_ topics to `3`.  So that even if two brokers would become unavailable at the same time, the system would still function. 
```

> We also recommend running with at least three zookeepers and two Schema registry servers.

Reviewer: ```suggestion
We also recommend running at least three zookeepers and two instances of the Confluent Schema registry.
```

> If deployed on kubernetes, consider running Kafka and Zookeeper on a StatefulSet workload, for persistent storage. The Schema registry can be run as a Deployment, as it is stateless.

Reviewer: ```suggestion
If you decide to run the Kafka cluster on Kubernetes, we recommendr unning Kafka and Zookeeper as StatefulSet workloads, for persistent storage. The Confluent Schema registry can be run as a Deployment, as it is stateless.
```

Apart from running your own kafka cluster, a managed version can also be used.

> The kafka cluster is usually started in the following order:

Reviewer: ```suggestion
The kafka cluster is usually started in the following order:

```

> - Start the `Zookeeper` nodes. The zookeeper nodes need to have the ports 2888 and 3888 open between them and the port 2181 open to the kafka servers. Note the hostnames of the zookeeper servers, which will be needed when configuring the kafka brokers.

Reviewer: ```suggestion
- Start the `ZooKeeper` nodes. The ZooKeeper nodes need to have the ports 2888 and 3888 open between them and the port 2181 open to the Kafka servers. Note down the hostnames of the zookeeper servers because they are required by the configuration of the Kafka brokers.
```

- Start and connect the `Kafka` nodes. The kafka nodes need the configuration how to connect to the zookeeper nodes, through the configuration variable `zookeeper.connect`, specified in the `/etc/kafka/server.properties` file. 

> - After kafka and Zookeeper are up and running, the Schema registry should be started, with the setting of how to reach the kafka brokers, usually through the configured parameter `kafkastore.bootstrap.server`, in the configuration file `/etc/schema-registry/schema-registry.properties`.

Reviewer: ```suggestion
- Once Kafka and ZooKeeper are up and running, the confluent Schema registry can be started. It requires Kafka brokers hostnames,  usually through the parameter `kafkastore.bootstrap.server` of the configuration file `/etc/schema-registry/schema-registry.properties`.
```

> The location of the configuration files, can vary depending on the particular installation.

Reviewer: ```suggestion
The location of the configuration files can vary depending on the particular installation.
```

> After configuring the Kafka cluster, few topics must be created, which are required by the `Airy apps`. Those topics can be found in the bash script `infrastructure/scripts/provision/create-topics.sh`. The script can be run, after setting properly the environment variables for the particular Kafka installation: 

Reviewer: ```suggestion
Once  the Kafka cluster is up and running, the required topics must be created. You can find them in the Bash script `infrastructure/scripts/provision/create-topics.sh`. The script requires the following environment variables to run: 
```

- `ZOOKEEPER` (default: airy-cp-zookeeper:2181)
- `PARTITIONS` (default: 10)
- `REPLICAS` (default: 1)

> Deploying Kafka on docker is not recommended for production environments with multiple brokers and under heavy load. However, we have a provided a way to deploy the whole kafka cluster on top of kubernetes with helm.

Reviewer: ```suggestion
We do not recommended running Kafka on docker for production environments. However, we provide a way to deploy the whole Kafka cluster on top of Kubernetes with Helm as we use this approach for local installations.
```

> To deploy kafka on kubernetes with helm, after setting the proper kubernetes config file and context, you must run:

Reviewer: ```suggestion
To deploy Kafka on Kubernetes with Helm, you can run:
```

```sh
helm install airy infrastructure/helm-chart/charts/kafka/
```

> By default, the `Schema registry` deployment is created with `replicas=0` because we want to ensure Zookeeper and Kafka are running properly beforehand, to avoid errors and crashing of this app.

Reviewer: ```suggestion
By default, the `Confluent Schema registry` deployment is created with `replicas=0`. We want to ensure ZooKeeper and Kafka are running properly before we start the service so we can avoid errors.
```

> You can customize the deployment, by editing the configuration files in `infrastructure/helm-chart/charts/kafka/charts`.

Reviewer: ```suggestion
To customize the deployment, you can edit the configuration files in `infrastructure/helm-chart/charts/kafka/charts`.
```

### PostgreSQL

A cluster of PostgreSQL servers can be used or a managed one such as AWS RDS.
If deployed on kubernetes, consider running the app on a StatefulSet workload, for persistent storage.

> After the server has been provisioned, a database must be created with proper credentials (username and password), required later when starting the `Airy apps`.

Reviewer: I think we should say more about what "proper" means here

> We have provided a helm chart, which can deploy PostgreSQL server in kubernetes, after setting the proper kubernetes config file and context. For creating the PostgreSQL server with helm, set the `pgPassword` variable in `infrastructure/helm-chart/charts/postgres/values.yaml` file and run:

Reviewer: ```suggestion
We provide a Helm chart to deploy a PostgreSQL server Kubernetes.  Set the `pgPassword` variable in `infrastructure/helm-chart/charts/postgres/values.yaml` file and run:
```

```sh
helm install postgres infrastructure/helm-chart/charts/postgres/
```

### Redis

> A cluster of redis servers can be used or a managed one such as AWS ElastiCache for Redis

Reviewer: ```suggestion
A cluster of redis servers can be used or a managed one such as AWS ElastiCache for Redis.
```

If deployed on kubernetes, consider running the app on a StatefulSet workload, for persistent storage.

> We have provided a helm chart, which can deploy the Redis server in kubernetes, after setting the proper kubernetes config file and context. For creating the redis cluster with helm run:

Reviewer: ```suggestion
We provided a Helm chart for a Redis cluster as well. You can run:
```

```sh
helm install redis infrastructure/helm-chart/charts/redis/
```

### Kubernetes or docker runtime

> The `Airy apps` run on top of kubernetes, which is our preferred runtime. All of the apps can also be deployed using docker, however we recommend kubernetes for more robust and reliable deployment in production environments.

Reviewer: as noted in other parts of the document, we have to be more explicit about this. If it's a requirement (and probably at the moment it should be so it simplifies getting started with the platform), we should say clearly

Author: I suggest we `recommend` and state that this guide is about deploying on `Kubernetes`. However, it is fully possible to deploy this also on docker.

In order to proceed with deploying the apps, we assume that you have a running kubernetes cluster, properly configured KUBECONF file and properly set context.

At the moment, the Airy Core Platform is fully tested on K3s, Minikube and AWS EKS. All of the following configuration and deployment instructions are applicable for all kubernetes implementations in general.

### Configuration

> After the underlying services are deployed, the Airy apps need to be created as docker containers on any docker runtime or inside kubernetes. Regardless of the deployment decision, they must have network access to the underlying services through the following variables which need to be defined before starting the containers.

Reviewer: ```suggestion
After the underlying services are deployed, the `Airy apps` need to be created as docker containers on any docker runtime or inside kubernetes. They must have network access to the underlying services through the following variables which need to be defined before starting the containers.
```

Connecting the `Airy apps` to the Kafka cluster, PostgreSQL and Redis can be done by creating a configuration file, prior to deploying the apps.

> Example with all possible configuration items, can be found in `infrastructure/airy.conf.all`. This file should be copied to `airy.conf` and edited according to the particular environment.

Reviewer: ```suggestion
An example with all possible configuration parameters can be found in `infrastructure/airy.conf.all`. This file should be copied to `airy.conf` and edited according to the particular environment.
```

### Deployment of the Airy apps

> The Airy apps can run on a docker runtime engine, however we recommend deploying the Airy Core Platform inside Kubernetes. For bootstrapping the entire Airy Core Platform on a single machine, we included a Vagrant configuration, inside the `infrastructure` directory, which creates a K3s Kubernetes implementation, running on top of Alpine Linux. All of the underlying services and the Airy apps run inside this Kubernetes box, however, they can also be deployed in a different manner.

Reviewer: this seems relevant to the local doc but now this one right?

Author: Will reword and keep a short paragraph. Will also put some of this content in the local doc.

> To deploy all of the `Airy apps` we have provided a helm chart, but first the `airy.conf` confiugration file needs to be edited in accordance to the environment.

Reviewer: ```suggestion
We provided a Helm chart to deploy the `Airy apps`  but, before you can run helm, you must configure the system via the `airy.conf` file.
```

```sh
cd infrastructure
cp airy.conf.all airy.conf
```

After editing the file, run:
```sh
cp airy.conf ./helm-chart/charts/apps/values.yaml 
helm install airy-apps ./helm-chart/charts/apps/ --timeout 1000s
```

> By default, all of the `Airy core` deployments start with `replicas=0`. To scale them up, run:

Reviewer: ```suggestion
By default,  the `Airy core` deployments start with `replicas=0` so to to scale them up, run:
```

```sh
kubectl scale deployment -l airy=api --replicas=1
kubectl scale deployment -l airy=sources --replicas=1
kubectl scale deployment -l airy=webhook --replicas=1
kubectl scale deployment -l airy=frontend --replicas=1
```

> Then you should have a running `Airy Core Platform` in your environment.

Reviewer: ```suggestion
At this point you should have a running `Airy Core Platform` in your environment 🎉 .
```

---

> title: Running the Airy Core Plaform in production

Reviewer 3: ```suggestion
title: Running the Airy Core Platform in production
```

sidebar_label: Production
---

This document provides an overview of the architecture of the Airy Core Platform and good practices for production environments. 

To better understand the architecture of the Airy Core Platform, please refer to the [Architecture](architecture.md) document.

## Requirements 

The `Airy apps` must run as separated docker containers and require access to several services which need to be in place for the system to be functional:

> - `Kafka cluster`: Kafka, Zookeeper and the Confluent Schema registry. These three services comprise the Kafka store. They are the default storage system of the Airy Core Platform. Kafka requires Zookeeper to  work. The Confluent Schema registry facilitates Avro typed data pipe-lines. All our Kafka based applications require the registry to work.

Reviewer 3: ```suggestion
- `Kafka cluster`: Kafka, Zookeeper and the Confluent Schema registry. These three services comprise the Kafka store. They are the default storage system of the Airy Core Platform. Kafka requires Zookeeper to  work. The Confluent Schema registry facilitates Avro typed data pipelines. All our Kafka based applications require the registry to work.
```

- Start the `ZooKeeper` nodes. The ZooKeeper nodes need to have the ports 2888 and 3888 open between them and the port 2181 open to the Kafka servers. Note down the hostnames of the zookeeper servers because they are required for the configuration of the Kafka brokers.
- Start and connect the `Kafka` nodes. The kafka nodes need the configuration how to connect to the zookeeper nodes, through the configuration variable `zookeeper.connect`, specified in the `/etc/kafka/server.properties` file. 
- Once Kafka and ZooKeeper are up and running, the confluent Schema registry can be started. It requires Kafka brokers hostnames,  usually through the parameter `kafkastore.bootstrap.server` of the configuration file `/etc/schema-registry/schema-registry.properties`.

The location of the configuration files can vary depending on the particular installation.

Once  the Kafka cluster is up and running, the required topics must be created. You can find them in the Bash script `infrastructure/scripts/provision/create-topics.sh`. The script requires the following environment variables to run: 
- `ZOOKEEPER` (default: airy-cp-zookeeper:2181)
- `PARTITIONS` (default: 10)
- `REPLICAS` (default: 1)

> We do not recommended running Kafka on docker for production environments. However, we provide a way to deploy the whole Kafka cluster on top of Kubernetes with Helm as we use this approach for local installations.

Reviewer 3: ```suggestion
We do not recommend running Kafka on docker for production environments. However, we provide a way to deploy the whole Kafka cluster on top of Kubernetes with Helm as we use this approach for local installations.
```

```sh
helm install postgres infrastructure/helm-chart/charts/postgres/
```

### Redis

A cluster of redis servers can be used or a managed one such as AWS ElastiCache for Redis.
If deployed on kubernetes, consider running the app on a StatefulSet workload, for persistent storage.

We provided a Helm chart for a Redis cluster as well. You can run:
```sh
helm install redis infrastructure/helm-chart/charts/redis/
```

### Kubernetes

The `Airy apps` run on top of Kubernetes which is our preferred runtime. We recommend Kubernetes for more robust and reliable deployment in production environments. However, the `Airy apps` can also be deployed using plain Docker runtime.

If you prefer to deploy the apps on a Docker runtime, you should start by inspecting the kubernetes manifests, where all the docker images and configuration parameters are specified. In order to generate the kubernetes manifests, run:
```sh
helm template ./infrastructure/helm-chart
```

The remaining of this guide describes how to deploy the `Airy Core Platform` on Kubernetes.

> In order to proceed with deploying the apps, we assume that you have a running kubernetes cluster, properly configured KUBECONF file and properly set context.

Author: Please have another look. What is in the first sentence, I think we should say somewhere in the document. Also the second sentence - that is actually a requirement for the apps to be deployed.

At the moment, the Airy Core Platform is fully tested on K3s, Minikube and AWS EKS. All of the following configuration and deployment instructions are applicable for all kubernetes implementations in general.

## Running the Airy apps

> Depending on implementation of the user, different choices can be made regarding the implementation of the underlying services and the whole system in general. As first pre-requisites, the underlying services are needed to be running. They can be implemented as separate docker containers, in different Kubernetes clusters, on a separate virtual machines or physical servers. Here is a brief description on the implementation and how to fully achieve high availability.

Reviewer 3: ```suggestion
Depending on implementation of the user, different choices can be made regarding the implementation of the underlying services and the whole system in general. As first pre-requisites, the underlying services are needed to be running. They can be implemented as separate docker containers, in different Kubernetes clusters, on separate virtual machines or physical servers. Here is a brief description on the implementation and how to fully achieve high availability.
```

### Deployment

We provided a Helm chart to deploy all of the `Airy apps`. Before you can run helm, you must configure the system via the `airy.conf` file.

```sh
cd infrastructure
cp airy.conf.all airy.conf
```

After editing the file, run:
```sh
cp airy.conf ./helm-chart/charts/apps/values.yaml 
helm install airy-apps ./helm-chart/charts/apps/ --timeout 1000s
```

By default,  the `Airy core` deployments start with `replicas=0` so to scale them up, run:
```sh
kubectl scale deployment -l airy=api --replicas=1
kubectl scale deployment -l airy=sources --replicas=1
kubectl scale deployment -l airy=webhook --replicas=1
kubectl scale deployment -l airy=frontend --replicas=1
```

At this point you should have a running `Airy Core Platform` in your environment 🎉.

To deploy with a different `image tag` (for example `beta` from the `develop` branch), you can run:
```sh
export APP_IMAGE_TAG=beta
helm install airy-apps ./helm-chart/charts/apps/ --set global.appImageTag=${APP_IMAGE_TAG} --timeout 1000s
```

## Network

### Sources

> As the Airy Core Platform processes messages from public sources, the `sources-SOURCE_NAME-webhook` pods should be exposed publicly on the Internet. Inside the helm chart and for the purposes of bootstrapping the whole platform you a local machine, we have included a [Ngrok](https://ngrok.com/) client, as a sidecar container in all of the `sources-SOURCE_NAME-webhook` pods. Ngrok is an open source reverse proxy, which creates a secure tunnel from a public endpoint to a local service. The Ngrok client connects to a Ngrok server which has public access to the internet and then provides a reversed proxy connectivity back to the webhook services, running in side the Kubernetes cluster.

Reviewer 3: ```suggestion
As the Airy Core Platform processes messages from public sources, the `sources-SOURCE_NAME-webhook` pods should be exposed publicly on the Internet. Inside the helm chart and for the purposes of bootstrapping the whole platform you a local machine, we have included a [Ngrok](https://ngrok.com/) client, as a sidecar container in all of the `sources-SOURCE_NAME-webhook` pods. Ngrok is an open source reverse proxy, which creates a secure tunnel from a public endpoint to a local service. The Ngrok client connects to a Ngrok server which has public access to the internet and then provides a reversed proxy connectivity back to the webhook services, running inside the Kubernetes cluster.
```

If you are not using Ngrok to expose the webhook services, you need to find alternative ways to expose the port `8080` of the `sources-SOURCE_NAME-webhook` pods, in accordance with your infrastructure. If your Kubernetes cluster has public access, you can also do this with Kubernetes ingress resources.

### Services

Inside the Kubernetes cluster, the helm chart installs the API services (with prefix `api-`) and the frontend services (with prefix `frontend-`). To access these services outside the Kubernetes cluster they need to be exposed with Ingress resources. You can choose an [Kubernetes ingress controller](https://kubernetes.io/docs/concepts/services-networking/ingress-controllers/) in accordance to your needs or preferences. If you are using the [Traefik](https://traefik.io/) ingress controller, you can edit the `infrastructure/network/ingress.yaml` file to modify the `host` records and directly apply it to your Kubernetes cluster.

```sh
kubectl apply -f infrastructure/network/ingress.yaml
```

You need to set different `host` attributes for the:
- API endpoints (defaults to `api.airy`)
- Webapp UI (defaults to `demo.airy`)
- Chat plugin (defaults to `chatplugin.airy`)

> If you are not using Traefik, you can use the `infrastructure/network/ingress.yaml` file as a guide to create your own kubernetes manifest for your preferred ingress controller.

Reviewer 3: ```suggestion
If you are not using Traefik, you can use the `infrastructure/network/ingress.yaml` file as a guide to creating your own kubernetes manifest for your preferred ingress controller.
```

Reviewer: the suggested change is not correct in English

---
title: Running the Airy Core Platform in production
sidebar_label: Production
---

This document provides our recommendations on how to run the Airy Core Platform in production environments. If you are not familiar with the architecture of the system, we suggest you read the [Architecture](architecture.md) document before proceeding.

## Requirements 

> The `Airy apps` are the internal and exposed services, which comprise the `Airy Core Platform`. They run as docker containers and require access to several base services to be in place before they can be started:

Reviewer: ```suggestion
The `Airy apps` are the services which comprise the `Airy Core Platform`. They run as docker containers and require access to several other services to be in place before they can be started:
```

### Configuration

After the [required services](#requirements) are deployed, you're ready to start the `Airy apps` inside a Kubernetes cluster. Connecting the `Airy apps` to the Kafka cluster, PostgreSQL and Redis can be done by creating a configuration file, prior to deploying the apps. Make sure that the `Airy apps` also have network connectivity to the required services.

The file `infrastructure/airy.conf.all` contains an example of all possible configuration parameters. This file should be copied to `airy.conf` and edited according to your environment:

```sh
cd infrastructure
cp airy.conf.all airy.conf
```

Edit the file to configure connections to the base services. Make sure that the following sections are configured correctly, so that the `Airy apps` to start properly:

```
apps:
  kafka:
    ...
  redis:
    ...
  postgresql:
    ...
```

### Deployment

> We provided a Helm chart to deploy all of the `Airy apps`. Before you can run helm, you must configure the system via the `airy.conf` file. After that run:

Reviewer: ```suggestion
We provided a Helm chart to deploy the `Airy apps`. Before you can run helm, you must configure the system via the `airy.conf` file, then you can proceed:
```

- `PostgreSQL`: Where we store authentication data.
- `Redis`: The queuing system used by our webhook system relies on Redis.

### Kafka cluster

The Kafka store requires multiple Kafka brokers, zookeepers, and Confluent Schema registry servers. We recommend using at least five Kafka brokers and  set the replication factor of _all_ topics to `3`.  So that even if two brokers would become unavailable at the same time, the system would still function. 

We also recommend running at least three zookeepers and two instances of the Confluent Schema registry.

If you decide to run the Kafka cluster on Kubernetes, we recommend running Kafka and Zookeeper as StatefulSet workloads, for persistent storage. The Confluent Schema registry can be run as a Deployment, as it is stateless.

Apart from running your own kafka cluster, a managed version can also be used.

The kafka cluster is usually started in the following order:

> - Start the `ZooKeeper` nodes. The ZooKeeper nodes need to have the ports 2888 and 3888 open between them and the port 2181 open to the Kafka servers. Note down the hostnames of the zookeeper servers because they are required for the configuration of the Kafka brokers.

Reviewer: ```suggestion
- Start the `ZooKeeper` nodes. The ZooKeeper nodes need to have the ports `2888` and `3888` open between them and the port `2181` open to the Kafka servers. Note down the hostnames of the ZooKeeper servers because they are required for the configuration of the Kafka brokers.
```

- Start and connect the `Kafka` nodes. The kafka nodes need the configuration how to connect to the zookeeper nodes, through the configuration variable `zookeeper.connect`, specified in the `/etc/kafka/server.properties` file. 
- Once Kafka and ZooKeeper are up and running, the confluent Schema registry can be started. It requires Kafka brokers hostnames,  usually through the parameter `kafkastore.bootstrap.server` of the configuration file `/etc/schema-registry/schema-registry.properties`.

The location of the configuration files can vary depending on the particular installation.

Once  the Kafka cluster is up and running, the required topics must be created. You can find them in the Bash script `infrastructure/scripts/provision/create-topics.sh`. The script requires the following environment variables to run: 
- `ZOOKEEPER` (default: airy-cp-zookeeper:2181)
- `PARTITIONS` (default: 10)
- `REPLICAS` (default: 1)

> We do not recommend running Kafka on docker for production environments. However, we provide a way to deploy the whole Kafka cluster on top of Kubernetes with Helm as we use this approach for local installations.

Reviewer: ```suggestion
We do not recommend running Kafka on docker for production environments. However, we provide a way to deploy the whole Kafka cluster on top of Kubernetes with Helm as we use this approach for test installations.
```

To deploy Kafka on Kubernetes with Helm, you can run:
```sh
helm install airy infrastructure/helm-chart/charts/kafka/
```

By default, the `Confluent Schema registry` deployment is created with `replicas=0`. We want to ensure ZooKeeper and Kafka are running properly before we start the service so we can avoid errors.

To customize the deployment, you can edit the configuration files in `infrastructure/helm-chart/charts/kafka/charts`.

### PostgreSQL

A cluster of PostgreSQL servers can be used or a managed one such as AWS RDS.
If deployed on kubernetes, consider running the app on a StatefulSet workload, for persistent storage.

> After the server has been provisioned, a database must be created, along with username and password which have full privileges on the database. These parameters will be required later when starting the `Airy apps`.

Reviewer: ```suggestion
After the server has been provisioned, a database must be created, along with username and password which have full privileges on the database. Note down these parameters as they are required to start the `Airy apps`.
```

We provide a Helm chart to deploy a PostgreSQL server Kubernetes.  Set the `pgPassword` variable in `infrastructure/helm-chart/charts/postgres/values.yaml` file and run:
```sh
helm install postgres infrastructure/helm-chart/charts/postgres/
```

### Redis

A cluster of redis servers can be used or a managed one such as AWS ElastiCache for Redis.
If deployed on kubernetes, consider running the app on a StatefulSet workload, for persistent storage.

We provided a Helm chart for a Redis cluster as well. You can run:
```sh
helm install redis infrastructure/helm-chart/charts/redis/
```

### Kubernetes

The `Airy apps` run on top of Kubernetes which is our preferred runtime. We recommend Kubernetes for more robust and reliable deployment in production environments. However, the `Airy apps` do not depend on Kubernetes in any way.

> If you prefer to deploy the `Airy apps` on a Docker runtime, you should start by inspecting the Kubernetes manifests, where all the Docker images and configuration parameters are specified. In order to generate the Kubernetes manifests, run:

Reviewer: ```suggestion
If you prefer to deploy the `Airy apps` on a Docker runtime, you can refer to our the Kubernetes manifests, where all the Docker images and configuration parameters are specified. In order to generate the Kubernetes manifests, run:
```

```sh
helm template ./infrastructure/helm-chart
```

The remaining of this guide describes how to deploy the `Airy Core Platform` on Kubernetes.

In order to proceed with deploying the apps, we assume that you have a running kubernetes cluster, properly configured KUBECONF file and properly set context.

## Running the Airy apps

So far the Airy Core Platform has been tested on K3s, Minikube and AWS EKS. The following configuration and deployment instructions are applicable to any Kubernetes implementation as they depend on widely supported Kubernetes features.

Depending on implementation of the user, different choices can be made regarding the implementation of the base services and the whole system in general. As first pre-requisites, the base services are needed to be running. They can be implemented as separate docker containers, in different Kubernetes clusters, on separate virtual machines or physical servers. Here is a brief description on the implementation and how to fully achieve high availability.

### Configuration

> After the [required services](#requirements) are deployed, you're ready to start the `Airy apps` inside a Kubernetes. Connecting the `Airy apps` to the Kafka cluster, PostgreSQL and Redis can be done by creating a configuration file, prior to deploying the apps. Make sure that the `Airy apps` also have network connectivity to the required services.

Reviewer: ```suggestion
After the [required services](#requirements) are deployed, you're ready to start the `Airy apps` inside a Kubernetes cluster. Connecting the `Airy apps` to the Kafka cluster, PostgreSQL and Redis can be done by creating a configuration file, prior to deploying the apps. Make sure that the `Airy apps` also have network connectivity to the required services.
```

> The file `infrastructure/airy.conf.all` contains an example of all possible configuration parameters. This file should be copied to `airy.conf` and edited according to the particular environment.

Reviewer: ```suggestion
The file `infrastructure/airy.conf.all` contains an example of all possible configuration parameters. This file should be copied to `airy.conf` and edited according to your environment:
```

    ...
```

### Deployment

We provided a Helm chart to deploy all of the `Airy apps`. Before you can run helm, you must configure the system via the `airy.conf` file. After that run:

```sh
cp airy.conf ./helm-chart/charts/apps/values.yaml 
helm install airy-apps ./helm-chart/charts/apps/ --timeout 1000s
```

By default, the `Airy apps` deployments start with `replicas=0` so to scale them up, run:
```sh
kubectl scale deployment -l airy=api --replicas=1
kubectl scale deployment -l airy=sources --replicas=1
kubectl scale deployment -l airy=webhook --replicas=1
kubectl scale deployment -l airy=frontend --replicas=1
```

At this point you should have a running `Airy Core Platform` in your environment 🎉.

To deploy with a different `image tag` (for example `beta` from the `develop` branch), you can run:
```sh
export AIRY_VERSION=beta
helm install airy-apps ./helm-chart/charts/apps/ --set global.appImageTag=${AIRY_VERSION} --timeout 1000s
```

If afterwards you need to modify or add other config parameters in the `airy.conf` file, after editing the file run:

```sh
cp airy.conf ./helm-chart/charts/apps/values.yaml
helm upgrade airy-apps ./helm-chart/charts/apps/ --timeout 1000s
```

> If you have deployed the Airy Core Platform with a specific version tag, export the `AIRY_VERSION` variable, before running `helm upgrade`:

Reviewer: ```suggestion
If you deploy the Airy Core Platform with a specific version tag, you must export the `AIRY_VERSION` variable before running `helm upgrade`:
```

```sh
cp airy.conf ./helm-chart/charts/apps/values.yaml
export AIRY_VERSION=beta
helm upgrade airy-apps ./helm-chart/charts/apps/ --set global.appImageTag=${AIRY_VERSION} --timeout 1000s
```

## Network

### Sources

> The helm chart creates separate NodePort service for every source with the naming convention  `sources-SOURCE_NAME-webhook`. These services need to be exposed publicly on the Internet, so that the sources can send messages and events to the webhook services, inside the kubernetes cluster.

Reviewer: ```suggestion
The helm chart creates separate NodePort service for every source with the naming convention  `sources-SOURCE_NAME-webhook`. These services must be exposed publicly on the Internet, so that the sources can send messages and events to the webhook services, inside the Kubernetes cluster.
```

```sh
kubectl get service -l airy=sources.webhook
```
The public endpoint will be in the `EXTERNAL-IP` column.

### Services

Inside the Kubernetes cluster, the helm chart installs the API services (with prefix `api-`) and the frontend services (with prefix `frontend-`). To access these services outside the Kubernetes cluster they need to be exposed with Ingress resources. You can choose an [Kubernetes ingress controller](https://kubernetes.io/docs/concepts/services-networking/ingress-controllers/) in accordance to your needs or preferences. If you are using the [Traefik](https://traefik.io/) ingress controller, you can edit the `infrastructure/network/ingress.yaml` file to modify the `host` records and directly apply it to your Kubernetes cluster.

```sh
kubectl apply -f infrastructure/network/ingress.yaml
```

You must set different `host` attributes for the following:

- API endpoints (defaults to `api.airy`)
- Demo (defaults to `demo.airy`)
- Chat plugin (defaults to `chatplugin.airy`)

If you are not using Traefik, you can use the `infrastructure/network/ingress.yaml` file as a guide to create your own kubernetes manifest for your preferred ingress controller.

> If your kubernetes cluster isn't directly reachable on the Internet, you will need a public LoadBalancer or `reverse proxy` to tunnel the traffic to the ports exposed by the ingress controller (usually ports 80 and 443).

Reviewer: ```suggestion
If your Kubernetes cluster is not directly reachable on the Internet, you will need a public LoadBalancer or a `reverse proxy` to tunnel the traffic to the ports exposed by the ingress controller (usually ports `80` and `443`).
```
